messaging/src/main/java/org/springframework/security/messaging/context/SecurityContextChannelInterceptor.java

```diff
@@ -118,14 +141,19 @@ public void setSecurityContextHolderStrategy(SecurityContextHolderStrategy strat
 
 	private void setup(Message<?> message) {
 		SecurityContext currentContext = this.securityContextHolderStrategy.getContext();
+		Object user = message.getHeaders().get(this.authenticationHeaderName);
+		Authentication authentication = getAuthentication(user);
+
+		if (authentication.equals(currentContext.getAuthentication())) {
```

Comment:
I'm not sure this is a good idea. Wouldn't this cause `cleanup` to break? If `setup` is called three times and `cleanup` is subsequently called three times, if one of the stack pushes was skipped, then `cleanup` will restore incorrect values. It seems like this would create problems for the scope of each authentication change.